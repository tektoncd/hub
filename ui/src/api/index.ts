import axios from 'axios';
import { API_URL } from '../config/constants';
import { ICategory } from '../store/category';
import { IResource, IVersion } from '../store/resource';

export interface Api {
  categories(): Promise<ICategory>;
  resources(): Promise<IResource>;
  resourceVersion(resourceId: number): Promise<IVersion>;
  versionUpdate(versionId: number): Promise<IVersion>;
}

export class Hub implements Api {
  async categories() {
    try {
      return axios.get(`${API_URL}/categories`).then((response) => response.data);
    } catch (err) {
      return err.response;
    }
  }

  async resources() {
    try {
      return axios.get(`${API_URL}/resources`).then((response) => response.data);
    } catch (err) {
      return err.response;
    }
  }

  async resourceVersion(resourceId: number) {
    try {
      return axios
        .get(`${API_URL}/resource/${resourceId}/versions`)
        .then((response) => response.data);
    } catch (err) {
      return err.response;
    }
  }

  async versionUpdate(versionId: number) {
    try {
      return axios
        .get(`${API_URL}/resource/version/${versionId}`)
        .then((response) => response.data);
    } catch (err) {
      return err.response;
    }
  }
}
