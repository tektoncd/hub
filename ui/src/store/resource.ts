import { types, Instance } from 'mobx-state-tree';
import fuzzysort from 'fuzzysort';
import moment, { Moment } from 'moment';
import { flow, getEnv } from 'mobx-state-tree';
import { Tag, ICategoryStore, ITag } from './category';
import { Api } from '../api';
import { Catalog, CatalogStore } from './catalog';
import { Kind, KindStore } from './kind';
import { assert } from './utils';

export const updatedAt = types.custom<string, Moment>({
  name: 'momentDate',
  fromSnapshot(value: string): Moment {
    return moment(new Date(value));
  },
  toSnapshot(value: Moment): string {
    return value.fromNow();
  },
  isTargetType(v: string) {
    return moment.isMoment(v);
  },
  getValidationMessage(v: string) {
    if (moment.isMoment(v)) {
      return 'Invalid moment object';
    }
    return '';
  }
});

const Version = types.model('Version', {
  id: types.identifierNumber,
  version: types.string,
  displayName: types.optional(types.string, ''),
  description: types.optional(types.string, ''),
  minPipelinesVersion: types.optional(types.string, ''),
  rawURL: types.string,
  webURL: types.string,
  updatedAt: types.optional(updatedAt, '')
});

export const Resource = types
  .model('Resource', {
    id: types.number,
    name: types.identifier,
    catalog: types.reference(Catalog),
    kind: types.reference(Kind),
    latestVersion: types.reference(Version),
    displayVersion: types.reference(Version),
    tags: types.array(types.reference(Tag)), // ["1", "2"]
    rating: types.number,
    versions: types.array(types.reference(Version)),
    displayName: ''
  })
  .views((self) => ({
    get resourceName() {
      if (self.displayName !== '') {
        return self.displayName;
      }
      return self.name;
    },
    get webURL() {
      const index = self.displayVersion.webURL.lastIndexOf('/');
      return self.displayVersion.webURL.substring(0, index + 1);
    },
    get summary() {
      const description = self.displayVersion.description;
      const index = description.indexOf('\n');
      return description.substring(0, index) || description;
    },
    get detailDescription() {
      const description = self.displayVersion.description;
      const index = description.indexOf('\n');
      return index !== -1 ? description.substring(index + 1).trim() : '';
    },
    get installCommand() {
      return `kubectl apply -f ${self.displayVersion.rawURL}`;
    }
  }));

export type IResource = Instance<typeof Resource>;
export type IVersion = Instance<typeof Version>;

export enum SortByFields {
  Unknown = '',
  Name = 'Name',
  Rating = 'Rating'
}

export const ResourceStore = types
  .model('ResourceStore', {
    resources: types.map(Resource),
    versions: types.map(Version),
    catalogs: types.optional(CatalogStore, {}),
    kinds: types.optional(KindStore, {}),
    sortBy: types.optional(types.enumeration(Object.values(SortByFields)), SortByFields.Unknown),
    tags: types.optional(types.map(Tag), {}),
    search: '',
    err: '',
    isLoading: true
  })
  .views((self) => ({
    get items() {
      return Array.from(self.resources.values());
    },
    get api(): Api {
      return getEnv(self).api;
    },
    get categories(): ICategoryStore {
      return getEnv(self).categories;
    }
  }))

  .actions((self) => ({
    setLoading(l: boolean) {
      self.isLoading = l;
    },
    setSearch(text: string) {
      self.search = text;
    },
    add(item: IResource) {
      self.resources.put(item);
    },
    setSortBy(field: string) {
      const key: SortByFields = SortByFields[field as keyof typeof SortByFields];
      self.sortBy = key;
    }
  }))

  .actions((self) => ({
    versionInfo: flow(function* (resourceName: string) {
      try {
        self.setLoading(true);

        const { api } = self;
        const resource = self.resources.get(resourceName);
        assert(resource);

        const json = yield api.resourceVersion(resource.id);

        const versions: IVersion[] = json.data.versions.map((v: IVersion) => ({
          id: v.id,
          version: v.version,
          webURL: v.webURL,
          rawURL: v.rawURL
        }));

        versions.forEach((v: IVersion) => {
          if (!self.versions.has(String(v.id))) {
            self.versions.put(v);
            if (self.resources.has(resourceName)) {
              const resource = self.resources.get(resourceName);
              assert(resource);
              resource.versions.push(v.id);
            }
          }
        });
      } catch (err) {
        self.err = err.toString();
      }
      self.setLoading(false);
    }),

    versionUpdate: flow(function* (versionId: number) {
      try {
        self.setLoading(true);

        const { api } = self;
        const json = yield api.versionUpdate(versionId);

        const versionData = json.data;

        const version: IVersion = {
          id: versionData.id,
          version: versionData.version,
          displayName: versionData.displayName,
          description: versionData.description,
          minPipelinesVersion: versionData.minPipelinesVersion,
          webURL: versionData.webURL,
          rawURL: versionData.rawURL,
          updatedAt: versionData.updatedAt
        };

        self.versions.put(version);
      } catch (err) {
        self.err = err.toString();
      }
      self.setLoading(false);
    }),

    load: flow(function* () {
      try {
        self.setLoading(true);

        const { api } = self;
        const json = yield api.resources();

        const kinds: string[] = json.data.map((r: IResource) => r.kind);
        kinds.forEach((k) => self.kinds.add(k));

        json.data.forEach((r: IResource) => {
          self.catalogs.add(r.catalog);
          self.versions.put(r.latestVersion);
        });

        // adding the tags to the store - normalized
        const tags: ITag[] = json.data.flatMap((item: IResource) => item.tags);

        tags.forEach((t) => (t != null ? self.tags.put(t) : null));

        const resources: IResource[] = json.data.map((r: IResource) => ({
          id: r.id,
          name: r.name,
          catalog: r.catalog.id,
          kind: r.kind,
          latestVersion: r.latestVersion.id,
          displayVersion: r.latestVersion.id,
          tags: r.tags != null ? r.tags.map((tag: ITag) => tag.id) : [],
          rating: r.rating,
          versions: [],
          displayName: r.latestVersion.displayName
        }));

        resources.forEach((r: IResource) => {
          r.versions.push(r.latestVersion);
          self.add(r);
        });
      } catch (err) {
        self.err = err.toString();
      }
      self.setLoading(false);
    })
  }))

  .actions((self) => ({
    afterCreate() {
      self.load();
    },
    setDisplayVersion(resourceName: string, versionId: string) {
      const resource = self.resources.get(resourceName);
      assert(resource);
      const version = self.versions.get(versionId);
      assert(version);
      if (version.id !== resource.displayVersion.id) {
        resource.displayVersion = version;
        self.versionUpdate(version.id);
      }
    }
  }))

  .views((self) => ({
    get filteredResources() {
      const { resources, kinds, catalogs, categories, search, sortBy } = self;
      const { selectedTags } = categories;

      let filteredItems: IResource[] = [];
      resources.forEach((r: IResource) => {
        const matchesKind = kinds.selected.size === 0 || kinds.selected.has(r.kind.name);
        const matchesCatalogs = catalogs.selected.size === 0 || catalogs.selected.has(r.catalog.id);
        const matchesTags = selectedTags.size === 0 || r.tags.some((t) => selectedTags.has(t.id));

        if (matchesKind && matchesCatalogs && matchesTags) {
          filteredItems.push(r);
        }
      });

      if (search.trim() !== '') {
        filteredItems = fuzzysort
          .go(search, filteredItems, { keys: ['name', 'displayName'] })
          .map((resource: Fuzzysort.KeysResult<IResource>) => resource.obj);
      }

      switch (sortBy) {
        case SortByFields.Rating:
          return filteredItems.sort((first: IResource, second: IResource) =>
            first.rating < second.rating ? 1 : first.rating > second.rating ? -1 : 0
          );

        case SortByFields.Name:
          return filteredItems.sort((first: IResource, second: IResource) =>
            first.name > second.name ? 1 : first.name < second.name ? -1 : 0
          );

        default:
          return filteredItems;
      }
    }
  }));

export type IResourceStore = Instance<typeof ResourceStore>;
