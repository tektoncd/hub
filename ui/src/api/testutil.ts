import { Api } from './';
import * as fs from 'fs';
import { ICategory } from '../store/category';
import { IResource, IVersion } from '../store/resource';

export class FakeHub implements Api {
  dataDir: string;

  constructor(dataDir: string) {
    this.dataDir = dataDir;
  }

  async categories() {
    const data = `${this.dataDir}/categories.json`;
    const ret = () => JSON.parse(fs.readFileSync(data).toString());
    return new Promise<ICategory>((resolve) => {
      setTimeout(() => resolve(ret()), 1000);
    });
  }

  async resources() {
    const data = `${this.dataDir}/resources.json`;

    const ret = () => JSON.parse(fs.readFileSync(data).toString());
    return new Promise<IResource>((resolve) => {
      setTimeout(() => resolve(ret()), 1000);
    });
  }

  async resourceVersion(resourceId: number) {
    const data = `${this.dataDir}/resource_versions_${resourceId}.json`;

    const ret = () => JSON.parse(fs.readFileSync(data).toString());
    return new Promise<IVersion>((resolve) => {
      setTimeout(() => resolve(ret()), 1000);
    });
  }

  async versionUpdate(versionId: number) {
    const data = `${this.dataDir}/version_details_${versionId}.json`;

    const ret = () => JSON.parse(fs.readFileSync(data).toString());
    return new Promise<IVersion>((resolve) => {
      setTimeout(() => resolve(ret()), 1000);
    });
  }
}
