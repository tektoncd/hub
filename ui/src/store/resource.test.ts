import { ResourceStore, Resource, SortByFields } from './resource';
import { getSnapshot } from 'mobx-state-tree';
import { when } from 'mobx';
import { FakeHub } from '../api/testutil';
import { CategoryStore } from './category';
import { assert } from './utils';

const TESTDATA_DIR = `${__dirname}/testdata`;
const api = new FakeHub(TESTDATA_DIR);

describe('Store Object', () => {
  it('can create a resource object', () => {
    const store = Resource.create({
      id: 5,
      name: 'buildah',
      catalog: '1',
      kind: 'Task',
      latestVersion: 1,
      displayVersion: 1,
      tags: ['1'],
      rating: 5
    });

    expect(store.name).toBe('buildah');
  });
});

describe('Store functions', () => {
  it('creates a resource store', (done) => {
    const store = ResourceStore.create(
      {},
      {
        api,
        categories: CategoryStore.create({}, { api })
      }
    );

    expect(store.isLoading).toBe(true);

    when(
      () => !store.isLoading,
      () => {
        expect(store.resources.size).toBe(7);
        expect(getSnapshot(store.resources)).toMatchSnapshot();
        done();
      }
    );
  });

  it('creates a catalog store', (done) => {
    const store = ResourceStore.create(
      {},
      {
        api,
        categories: CategoryStore.create({}, { api })
      }
    );
    expect(store.isLoading).toBe(true);

    when(
      () => !store.isLoading,
      () => {
        expect(store.resources.size).toBe(7);
        expect(getSnapshot(store.catalogs)).toMatchSnapshot();
        done();
      }
    );
  });

  it('creates a kind store', (done) => {
    const store = ResourceStore.create(
      {},
      {
        api,
        categories: CategoryStore.create({}, { api })
      }
    );

    expect(store.isLoading).toBe(true);

    when(
      () => !store.isLoading,
      () => {
        expect(store.resources.size).toBe(7);
        expect(getSnapshot(store.kinds)).toMatchSnapshot();
        done();
      }
    );
  });

  it('filter resources based on selected catalog', (done) => {
    const store = ResourceStore.create(
      {},
      {
        api,
        categories: CategoryStore.create({}, { api })
      }
    );
    expect(store.isLoading).toBe(true);

    when(
      () => !store.isLoading,
      () => {
        expect(store.resources.size).toBe(7);
        const { items } = store.catalogs;

        const catalogs = items.get('2');
        assert(catalogs);
        catalogs.toggle();

        const filtered = store.filteredResources;
        expect(filtered.length).toBe(1);
        expect(filtered[0].name).toBe('hub');
        expect(filtered[0].catalog.name).toBe('tekton-hub');

        done();
      }
    );
  });

  it('filter resources based on selected kind', (done) => {
    const store = ResourceStore.create(
      {},
      {
        api,
        categories: CategoryStore.create({}, { api })
      }
    );
    expect(store.isLoading).toBe(true);

    when(
      () => !store.isLoading,
      () => {
        expect(store.resources.size).toBe(7);

        const kind = store.kinds.items.get('Pipeline');
        assert(kind);
        kind.toggle();

        expect(store.filteredResources.length).toBe(1);
        expect(store.filteredResources[0].name).toBe('hub');
        expect(store.filteredResources[0].kind.name).toBe('Pipeline');

        done();
      }
    );
  });

  it('filter resources based on selected kind and catalog', (done) => {
    const store = ResourceStore.create(
      {},
      {
        api,
        categories: CategoryStore.create({}, { api })
      }
    );
    expect(store.isLoading).toBe(true);

    when(
      () => !store.isLoading,
      () => {
        expect(store.resources.size).toBe(7);

        const kinds = store.kinds.items.get('Task');
        assert(kinds);

        const catalogs = store.catalogs.items.get('1');
        assert(catalogs);

        kinds.toggle();
        catalogs.toggle();

        store.setSearch('golang');

        store.filteredResources;

        expect(store.filteredResources.length).toBe(1);

        done();
      }
    );
  });

  it('makes sure to not add duplicate resources', (done) => {
    const store = ResourceStore.create(
      {},
      {
        api,
        categories: CategoryStore.create({}, { api })
      }
    );

    expect(store.isLoading).toBe(true);

    when(
      () => !store.isLoading,
      () => {
        const item = Resource.create({
          id: 44,
          name: 'golang-build',
          catalog: 1,
          kind: 'Task',
          latestVersion: 47,
          displayVersion: 47,
          tags: [1],
          rating: 5,
          versions: [47],
          displayName: 'golang build'
        });

        store.add(item);
        expect(store.resources.size).toBe(7);

        expect(getSnapshot(store.resources)).toMatchSnapshot();

        done();
      }
    );
  });

  it('it checks if the related date is a string', (done) => {
    const store = ResourceStore.create(
      {},
      {
        api,
        categories: CategoryStore.create({}, { api })
      }
    );
    expect(store.isLoading).toBe(true);
    when(
      () => !store.isLoading,
      () => {
        expect(store.isLoading).toBe(false);
        expect(store.resources.size).toBe(7);

        const versions = store.versions.get('1');
        assert(versions);
        expect(typeof versions.updatedAt.fromNow()).toBe('string');

        done();
      }
    );
  });

  it('should filter resources based on search', (done) => {
    const store = ResourceStore.create(
      {},
      {
        api,
        categories: CategoryStore.create({}, { api })
      }
    );
    expect(store.isLoading).toBe(true);
    when(
      () => !store.isLoading,
      () => {
        store.setSearch('golang');
        expect(store.filteredResources[0].name).toBe('golang-build');

        done();
      }
    );
  });

  it('it should return displayName', (done) => {
    const store = ResourceStore.create(
      {},
      {
        api,
        categories: CategoryStore.create({}, { api })
      }
    );
    expect(store.isLoading).toBe(true);
    when(
      () => !store.isLoading,
      () => {
        expect(store.isLoading).toBe(false);
        expect(store.resources.size).toBe(7);

        const resource = store.resources.get('aws-cli');
        assert(resource);

        expect(resource.resourceName).toBe('aws cli');
        done();
      }
    );
  });

  it('should sort resources based on selected key', (done) => {
    const store = ResourceStore.create(
      {},
      {
        api,
        categories: CategoryStore.create({}, { api })
      }
    );
    expect(store.isLoading).toBe(true);
    when(
      () => !store.isLoading,
      () => {
        const rating: SortByFields = SortByFields[SortByFields.Rating];
        store.setSortBy(rating);

        expect(store.filteredResources[0].rating).toBe(5);
        expect(store.filteredResources[0].name).toBe('aws-cli');

        const name: SortByFields = SortByFields[SortByFields.Name];
        store.setSortBy(name);

        expect(store.filteredResources[0].name).toBe('ansible-runner');

        done();
      }
    );
  });

  it('it should return webURL', (done) => {
    const store = ResourceStore.create(
      {},
      {
        api,
        categories: CategoryStore.create({}, { api })
      }
    );
    expect(store.isLoading).toBe(true);
    when(
      () => !store.isLoading,
      () => {
        expect(store.isLoading).toBe(false);
        expect(store.resources.size).toBe(7);

        const resource = store.resources.get('aws-cli');
        assert(resource);

        expect(resource.webURL).toBe(
          'https://github.com/tektoncd/catalog/tree/master/task/aws-cli/0.1/'
        );
        done();
      }
    );
  });

  it('it should return summary', (done) => {
    const store = ResourceStore.create(
      {},
      {
        api,
        categories: CategoryStore.create({}, { api })
      }
    );
    expect(store.isLoading).toBe(true);
    when(
      () => !store.isLoading,
      () => {
        expect(store.isLoading).toBe(false);
        expect(store.resources.size).toBe(7);

        const resource = store.resources.get('aws-cli');
        assert(resource);

        expect(resource.summary).toBe(
          'This task performs operations on Amazon Web Services resources using aws.'
        );
        done();
      }
    );
  });

  it('it should return detail description', (done) => {
    const store = ResourceStore.create(
      {},
      {
        api,
        categories: CategoryStore.create({}, { api })
      }
    );
    expect(store.isLoading).toBe(true);
    when(
      () => !store.isLoading,
      () => {
        expect(store.isLoading).toBe(false);
        expect(store.resources.size).toBe(7);

        const resource = store.resources.get('buildah');
        assert(resource);

        expect(resource.detailDescription).toBe(
          "Buildah Task builds source into a container image using Project Atomic's Buildah build tool.It uses Buildah's support for building from Dockerfiles, using its buildah bud command.This command executes the directives in the Dockerfile to assemble a container image, then pushes that image to a container registry."
        );
        done();
      }
    );
  });

  it('it should return install command', (done) => {
    const store = ResourceStore.create(
      {},
      {
        api,
        categories: CategoryStore.create({}, { api })
      }
    );
    expect(store.isLoading).toBe(true);
    when(
      () => !store.isLoading,
      () => {
        expect(store.isLoading).toBe(false);
        expect(store.resources.size).toBe(7);

        const resource = store.resources.get('aws-cli');
        assert(resource);

        expect(resource.installCommand).toBe(
          'kubectl apply -f https://raw.githubusercontent.com/tektoncd/catalog/master/task/aws-cli/0.1/aws-cli.yaml'
        );
        done();
      }
    );
  });

  it('update versions list for buildah resource', (done) => {
    const store = ResourceStore.create(
      {},
      {
        api,
        categories: CategoryStore.create({}, { api })
      }
    );

    expect(store.isLoading).toBe(true);

    when(
      () => !store.isLoading,
      () => {
        expect(store.resources.size).toBe(7);
        expect(getSnapshot(store.resources)).toMatchSnapshot();
        store.versionInfo('buildah');
        when(
          () => !store.isLoading,
          () => {
            const resource = store.resources.get('buildah');
            assert(resource);
            expect(resource.versions.length).toBe(2);
            done();
          }
        );
      }
    );
  });

  it('fetch 0.1 version details for buildah resource', (done) => {
    const store = ResourceStore.create(
      {},
      {
        api,
        categories: CategoryStore.create({}, { api })
      }
    );

    expect(store.isLoading).toBe(true);

    when(
      () => !store.isLoading,
      () => {
        expect(store.resources.size).toBe(7);
        expect(getSnapshot(store.resources)).toMatchSnapshot();
        store.versionInfo('buildah');
        when(
          () => !store.isLoading,
          () => {
            const resource = store.resources.get('buildah');
            assert(resource);
            expect(resource.versions.length).toBe(2);
            store.versionUpdate(13);
            when(
              () => !store.isLoading,
              () => {
                expect(resource.versions[1].minPipelinesVersion).toBe('0.12.1');
                expect(resource.versions[1].version).toBe('0.1');
                done();
              }
            );
          }
        );
      }
    );
  });

  it('set 0.1 as display version for buildah', (done) => {
    const store = ResourceStore.create(
      {},
      {
        api,
        categories: CategoryStore.create({}, { api })
      }
    );

    expect(store.isLoading).toBe(true);

    when(
      () => !store.isLoading,
      () => {
        expect(store.resources.size).toBe(7);
        expect(getSnapshot(store.resources)).toMatchSnapshot();
        store.versionInfo('buildah');
        when(
          () => !store.isLoading,
          () => {
            const resource = store.resources.get('buildah');
            assert(resource);
            expect(resource.versions.length).toBe(2);
            store.setDisplayVersion('buildah', '13');
            when(
              () => !store.isLoading,
              () => {
                expect(resource.versions[1].minPipelinesVersion).toBe('0.12.1');
                expect(resource.versions[1].version).toBe('0.1');
                done();
              }
            );
          }
        );
      }
    );
  });
});
