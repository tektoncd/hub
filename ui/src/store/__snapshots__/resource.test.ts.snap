// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store functions creates a catalog store 1`] = `
Object {
  "items": Object {
    "1": Object {
      "id": 1,
      "name": "tekton",
      "selected": false,
      "type": "community",
    },
    "2": Object {
      "id": 2,
      "name": "tekton-hub",
      "selected": false,
      "type": "official",
    },
  },
}
`;

exports[`Store functions creates a kind store 1`] = `
Object {
  "items": Object {
    "Pipeline": Object {
      "name": "Pipeline",
      "selected": false,
    },
    "Task": Object {
      "name": "Task",
      "selected": false,
    },
  },
}
`;

exports[`Store functions creates a resource store 1`] = `
Object {
  "ansible-runner": Object {
    "catalog": 1,
    "displayName": "Ansible Runner",
    "displayVersion": 1,
    "id": 1,
    "kind": "Task",
    "latestVersion": 1,
    "name": "ansible-runner",
    "rating": 4.5,
    "tags": Array [
      2,
    ],
    "versions": Array [
      1,
    ],
  },
  "aws-cli": Object {
    "catalog": 1,
    "displayName": "aws cli",
    "displayVersion": 4,
    "id": 4,
    "kind": "Task",
    "latestVersion": 4,
    "name": "aws-cli",
    "rating": 5,
    "tags": Array [
      2,
    ],
    "versions": Array [
      4,
    ],
  },
  "buildah": Object {
    "catalog": 1,
    "displayName": "",
    "displayVersion": 105,
    "id": 13,
    "kind": "Task",
    "latestVersion": 105,
    "name": "buildah",
    "rating": 4,
    "tags": Array [
      8,
    ],
    "versions": Array [
      105,
    ],
  },
  "golang-build": Object {
    "catalog": 1,
    "displayName": "golang build",
    "displayVersion": 47,
    "id": 44,
    "kind": "Task",
    "latestVersion": 47,
    "name": "golang-build",
    "rating": 5,
    "tags": Array [
      1,
    ],
    "versions": Array [
      47,
    ],
  },
  "hub": Object {
    "catalog": 2,
    "displayName": "Hub Pipeline",
    "displayVersion": 1,
    "id": 10,
    "kind": "Pipeline",
    "latestVersion": 1,
    "name": "hub",
    "rating": 4.5,
    "tags": Array [
      10,
    ],
    "versions": Array [
      1,
    ],
  },
  "jenkins": Object {
    "catalog": 1,
    "displayName": "jenkins operation",
    "displayVersion": 104,
    "id": 98,
    "kind": "Task",
    "latestVersion": 104,
    "name": "jenkins",
    "rating": 5,
    "tags": Array [
      57,
      56,
    ],
    "versions": Array [
      104,
    ],
  },
  "jib-maven": Object {
    "catalog": 1,
    "displayName": "jib maven",
    "displayVersion": 57,
    "id": 54,
    "kind": "Task",
    "latestVersion": 57,
    "name": "jib-maven",
    "rating": 5,
    "tags": Array [
      8,
    ],
    "versions": Array [
      57,
    ],
  },
}
`;

exports[`Store functions fetch 0.1 version details for buildah resource 1`] = `
Object {
  "ansible-runner": Object {
    "catalog": 1,
    "displayName": "Ansible Runner",
    "displayVersion": 1,
    "id": 1,
    "kind": "Task",
    "latestVersion": 1,
    "name": "ansible-runner",
    "rating": 4.5,
    "tags": Array [
      2,
    ],
    "versions": Array [
      1,
    ],
  },
  "aws-cli": Object {
    "catalog": 1,
    "displayName": "aws cli",
    "displayVersion": 4,
    "id": 4,
    "kind": "Task",
    "latestVersion": 4,
    "name": "aws-cli",
    "rating": 5,
    "tags": Array [
      2,
    ],
    "versions": Array [
      4,
    ],
  },
  "buildah": Object {
    "catalog": 1,
    "displayName": "",
    "displayVersion": 105,
    "id": 13,
    "kind": "Task",
    "latestVersion": 105,
    "name": "buildah",
    "rating": 4,
    "tags": Array [
      8,
    ],
    "versions": Array [
      105,
    ],
  },
  "golang-build": Object {
    "catalog": 1,
    "displayName": "golang build",
    "displayVersion": 47,
    "id": 44,
    "kind": "Task",
    "latestVersion": 47,
    "name": "golang-build",
    "rating": 5,
    "tags": Array [
      1,
    ],
    "versions": Array [
      47,
    ],
  },
  "hub": Object {
    "catalog": 2,
    "displayName": "Hub Pipeline",
    "displayVersion": 1,
    "id": 10,
    "kind": "Pipeline",
    "latestVersion": 1,
    "name": "hub",
    "rating": 4.5,
    "tags": Array [
      10,
    ],
    "versions": Array [
      1,
    ],
  },
  "jenkins": Object {
    "catalog": 1,
    "displayName": "jenkins operation",
    "displayVersion": 104,
    "id": 98,
    "kind": "Task",
    "latestVersion": 104,
    "name": "jenkins",
    "rating": 5,
    "tags": Array [
      57,
      56,
    ],
    "versions": Array [
      104,
    ],
  },
  "jib-maven": Object {
    "catalog": 1,
    "displayName": "jib maven",
    "displayVersion": 57,
    "id": 54,
    "kind": "Task",
    "latestVersion": 57,
    "name": "jib-maven",
    "rating": 5,
    "tags": Array [
      8,
    ],
    "versions": Array [
      57,
    ],
  },
}
`;

exports[`Store functions makes sure to not add duplicate resources 1`] = `
Object {
  "ansible-runner": Object {
    "catalog": 1,
    "displayName": "Ansible Runner",
    "displayVersion": 1,
    "id": 1,
    "kind": "Task",
    "latestVersion": 1,
    "name": "ansible-runner",
    "rating": 4.5,
    "tags": Array [
      2,
    ],
    "versions": Array [
      1,
    ],
  },
  "aws-cli": Object {
    "catalog": 1,
    "displayName": "aws cli",
    "displayVersion": 4,
    "id": 4,
    "kind": "Task",
    "latestVersion": 4,
    "name": "aws-cli",
    "rating": 5,
    "tags": Array [
      2,
    ],
    "versions": Array [
      4,
    ],
  },
  "buildah": Object {
    "catalog": 1,
    "displayName": "",
    "displayVersion": 105,
    "id": 13,
    "kind": "Task",
    "latestVersion": 105,
    "name": "buildah",
    "rating": 4,
    "tags": Array [
      8,
    ],
    "versions": Array [
      105,
    ],
  },
  "golang-build": Object {
    "catalog": 1,
    "displayName": "golang build",
    "displayVersion": 47,
    "id": 44,
    "kind": "Task",
    "latestVersion": 47,
    "name": "golang-build",
    "rating": 5,
    "tags": Array [
      1,
    ],
    "versions": Array [
      47,
    ],
  },
  "hub": Object {
    "catalog": 2,
    "displayName": "Hub Pipeline",
    "displayVersion": 1,
    "id": 10,
    "kind": "Pipeline",
    "latestVersion": 1,
    "name": "hub",
    "rating": 4.5,
    "tags": Array [
      10,
    ],
    "versions": Array [
      1,
    ],
  },
  "jenkins": Object {
    "catalog": 1,
    "displayName": "jenkins operation",
    "displayVersion": 104,
    "id": 98,
    "kind": "Task",
    "latestVersion": 104,
    "name": "jenkins",
    "rating": 5,
    "tags": Array [
      57,
      56,
    ],
    "versions": Array [
      104,
    ],
  },
  "jib-maven": Object {
    "catalog": 1,
    "displayName": "jib maven",
    "displayVersion": 57,
    "id": 54,
    "kind": "Task",
    "latestVersion": 57,
    "name": "jib-maven",
    "rating": 5,
    "tags": Array [
      8,
    ],
    "versions": Array [
      57,
    ],
  },
}
`;

exports[`Store functions set 0.1 as display version for buildah 1`] = `
Object {
  "ansible-runner": Object {
    "catalog": 1,
    "displayName": "Ansible Runner",
    "displayVersion": 1,
    "id": 1,
    "kind": "Task",
    "latestVersion": 1,
    "name": "ansible-runner",
    "rating": 4.5,
    "tags": Array [
      2,
    ],
    "versions": Array [
      1,
    ],
  },
  "aws-cli": Object {
    "catalog": 1,
    "displayName": "aws cli",
    "displayVersion": 4,
    "id": 4,
    "kind": "Task",
    "latestVersion": 4,
    "name": "aws-cli",
    "rating": 5,
    "tags": Array [
      2,
    ],
    "versions": Array [
      4,
    ],
  },
  "buildah": Object {
    "catalog": 1,
    "displayName": "",
    "displayVersion": 105,
    "id": 13,
    "kind": "Task",
    "latestVersion": 105,
    "name": "buildah",
    "rating": 4,
    "tags": Array [
      8,
    ],
    "versions": Array [
      105,
    ],
  },
  "golang-build": Object {
    "catalog": 1,
    "displayName": "golang build",
    "displayVersion": 47,
    "id": 44,
    "kind": "Task",
    "latestVersion": 47,
    "name": "golang-build",
    "rating": 5,
    "tags": Array [
      1,
    ],
    "versions": Array [
      47,
    ],
  },
  "hub": Object {
    "catalog": 2,
    "displayName": "Hub Pipeline",
    "displayVersion": 1,
    "id": 10,
    "kind": "Pipeline",
    "latestVersion": 1,
    "name": "hub",
    "rating": 4.5,
    "tags": Array [
      10,
    ],
    "versions": Array [
      1,
    ],
  },
  "jenkins": Object {
    "catalog": 1,
    "displayName": "jenkins operation",
    "displayVersion": 104,
    "id": 98,
    "kind": "Task",
    "latestVersion": 104,
    "name": "jenkins",
    "rating": 5,
    "tags": Array [
      57,
      56,
    ],
    "versions": Array [
      104,
    ],
  },
  "jib-maven": Object {
    "catalog": 1,
    "displayName": "jib maven",
    "displayVersion": 57,
    "id": 54,
    "kind": "Task",
    "latestVersion": 57,
    "name": "jib-maven",
    "rating": 5,
    "tags": Array [
      8,
    ],
    "versions": Array [
      57,
    ],
  },
}
`;

exports[`Store functions update versions list for buildah resource 1`] = `
Object {
  "ansible-runner": Object {
    "catalog": 1,
    "displayName": "Ansible Runner",
    "displayVersion": 1,
    "id": 1,
    "kind": "Task",
    "latestVersion": 1,
    "name": "ansible-runner",
    "rating": 4.5,
    "tags": Array [
      2,
    ],
    "versions": Array [
      1,
    ],
  },
  "aws-cli": Object {
    "catalog": 1,
    "displayName": "aws cli",
    "displayVersion": 4,
    "id": 4,
    "kind": "Task",
    "latestVersion": 4,
    "name": "aws-cli",
    "rating": 5,
    "tags": Array [
      2,
    ],
    "versions": Array [
      4,
    ],
  },
  "buildah": Object {
    "catalog": 1,
    "displayName": "",
    "displayVersion": 105,
    "id": 13,
    "kind": "Task",
    "latestVersion": 105,
    "name": "buildah",
    "rating": 4,
    "tags": Array [
      8,
    ],
    "versions": Array [
      105,
    ],
  },
  "golang-build": Object {
    "catalog": 1,
    "displayName": "golang build",
    "displayVersion": 47,
    "id": 44,
    "kind": "Task",
    "latestVersion": 47,
    "name": "golang-build",
    "rating": 5,
    "tags": Array [
      1,
    ],
    "versions": Array [
      47,
    ],
  },
  "hub": Object {
    "catalog": 2,
    "displayName": "Hub Pipeline",
    "displayVersion": 1,
    "id": 10,
    "kind": "Pipeline",
    "latestVersion": 1,
    "name": "hub",
    "rating": 4.5,
    "tags": Array [
      10,
    ],
    "versions": Array [
      1,
    ],
  },
  "jenkins": Object {
    "catalog": 1,
    "displayName": "jenkins operation",
    "displayVersion": 104,
    "id": 98,
    "kind": "Task",
    "latestVersion": 104,
    "name": "jenkins",
    "rating": 5,
    "tags": Array [
      57,
      56,
    ],
    "versions": Array [
      104,
    ],
  },
  "jib-maven": Object {
    "catalog": 1,
    "displayName": "jib maven",
    "displayVersion": 57,
    "id": 54,
    "kind": "Task",
    "latestVersion": 57,
    "name": "jib-maven",
    "rating": 5,
    "tags": Array [
      8,
    ],
    "versions": Array [
      57,
    ],
  },
}
`;
